patroni/validator.py

```diff
@@ -0,0 +1,117 @@
+            logger.info("postgresql.bin_dir is not defined")
+        else:
+            if os.path.exists(config["postgresql"]["bin_dir"]):
+                for program in ["pg_ctl", "initdb", "pg_controldata", "pg_basebackup"]:
+                    exe_file = os.path.join(config["postgresql"]["bin_dir"], program)
+                    if not (os.path.isfile(exe_file) and os.access(exe_file, os.X_OK)):
+                        logger.warning("Program '%s' not found.", program)
+            else:
+                logger.warning("Directory '%s' does not exist.", config['postgresql']['bin_dir'])
+                for program in ["pg_ctl", "initdb", "pg_controldata", "pg_basebackup"]:
+                    if not which(program):
+                        logger.warning("Program '%s' not found.", program)
+        if "data_dir" not in config["postgresql"]:
+            logger.info("postgresql.data_dir is not defined")
+        else:
+            if not data_directory_empty(config["postgresql"]["data_dir"]):
+                if not os.path.exists(
+                    os.path.join(config["postgresql"]["data_dir"], "PG_VERSION")
+                ):
+                    logger.warning(
+                        "data_dir doesn't look like a valid data directory make sure you provide a valid path in configuration"
+                    )
+                elif not os.path.isdir(
+                    os.path.join(config["postgresql"]["data_dir"], "pg_wal")
+                ):
+                    logger.warning(
+                        'data dir for the cluster is not empty, but doesn\'t contain "pg_wal" directory'
+                    )
+            data_dir = config["postgresql"]["data_dir"]
+            if "parameters" not in config["postgresql"]:
+                logger.info("postgresql.parameters is not defined")
+            else:
+                xlogdir = config["postgresql"]["parameters"].get(
+                    "xlogdir", os.path.join(data_dir, "pg_xlog")
+                )
+                waldir = config["postgresql"]["parameters"].get(
+                    "waldir", os.path.join(data_dir, "pg_wal")
+                )
+                log_directory = config["postgresql"]["parameters"].get(
+                    "log_directory", os.path.join(data_dir, "pg_log")
+                )
+                if waldir == log_directory:
+                    logger.warning(
+                        "waldir(%s) and log_directory (%s) are pointing to the same path", waldir, log_directory
+                    )
+                if xlogdir == log_directory:
+                    logger.warning(
+                        "xlogdir(%s) and log_directory (%s) are pointing to the same path", xlogdir, log_directory
+                    )
+
+        if "listen" not in config["postgresql"]:
+            logger.warning("postgresql.listen is not defined")
+        else:
+            host, port = split_host_port(config["postgresql"]["listen"], 5432)
```

Comment:
The `listen` could be in the format: `localhost,10.0.0.1:5433`

Reply:
There is also an optional `connect_address`.
If it is not defined, the value from `listen` will be used.
It is very important to check that the `host` in the `connect_adress` isn't a `localhost`, `127.0.0.1`, `0.0.0.0`, `::1`, `*` or something similar.